### PR TITLE
support for metadata store PasswordProvider interface

### PR DESCRIPTION
**Pull request by himanshug**
this patch enables ppl to write druid module extension which implements the necessary logic for getting the password for metadata store. default implementation provides the necessary backward compatibility to read the password in plain text from the druid configuration file.

internally, we have an extension which gets the password from a secure mechanism instead of having password stored in the configuration file.

Here is a reference implementation --
https://github.com/himanshug/druid-pwd-provider-extn-sample

### CODE REVIEW

File: common/src/main/java/io/druid/metadata/DefaultPasswordProvider.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  @JsonCreator
+  public DefaultPasswordProvider(Object spec)
+  {
+    password = convertValue(spec);
+  }
+
+  private static String convertValue(Object spec) {
+    if(spec instanceof String) {
+      return spec.toString();
+    } else if(spec instanceof Map && ((Map)spec).containsKey(PASSWORD_KEY)) {
```
Comment (xvrl) [inline common/src/main/java/io/druid/metadata/DefaultPasswordProvider.java]: is there a reason to support anything other than String here? Supporting both map and string seems redundant.

File: common/src/main/java/io/druid/metadata/DefaultPasswordProvider.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
+/*
```
Comment (xvrl) [inline common/src/main/java/io/druid/metadata/DefaultPasswordProvider.java]: Would you mind updating the license header to follow the template here? https://github.com/druid-io/druid/blob/master/processing/src/main/java/io/druid/segment/NullDimensionSelector.java#L1

This is in line with what the ASF does for their own projects http://www.apache.org/legal/src-headers.html#headers

File: common/src/main/java/io/druid/metadata/DefaultPasswordProvider.java (hunk 1 above)
Comment (himanshug) [inline common/src/main/java/io/druid/metadata/DefaultPasswordProvider.java]: supporting string is required for backward compatibility. For example, currently users would have their configuration like following...
druid.metadata.storage.connector.password=diurd
this will stop working if we removed String support.

Additionally, map allows the new way of specifying same password, for example like
druid.metadata.storage.connector.password={"type":"default","password":"diurd"}
this is not required strictly but nice for consistency and a ref on how it would be if you wrote an extension.

Comment (xvrl) [inline common/src/main/java/io/druid/metadata/DefaultPasswordProvider.java]: ah ok, I understood the need for string, but I was confused about the need for Map.

Since you already declare a `default` type as part of your jackson annotation, I don't see why you need to parse the map directly. You should be able to just add a `@JsonProperty("password") password` field to you constructor.

You can then add a static fromString method to a separate class and make that the `defaultImpl` for jackson.

Have a look at [QueryGranularity](https://github.com/druid-io/druid/blob/master/processing/src/main/java/io/druid/granularity/QueryGranularity.java#L46) where we do something similar, with a fromString method annotated with `@JsonCreator` that instantiates the default, but everything else using properly annotated constructors.

Comment (cheddar) [inline common/src/main/java/io/druid/metadata/DefaultPasswordProvider.java]: The stuff from a Map<> is required in order to read the JSON generated by Jackson when writing the thing out.  We could make Jackson write it out as just a String, but that would introduce issues with compatibility in the future if we needed to add extra fields.

File: common/src/main/java/io/druid/metadata/DefaultPasswordProvider.java (hunk 2 above)
Comment (cheddar) [inline common/src/main/java/io/druid/metadata/DefaultPasswordProvider.java]: Hrm, btw, it looks like that wasn't done for other files:

https://github.com/druid-io/druid/blob/master/common/src/main/java/io/druid/common/config/ConfigManager.java

Not sure when it is and isn't used.  We can definitely switch the header around, but I had told @himanshug to use this header because I hadn't noticed that you were using the other one yet...

Comment (xvrl) [inline common/src/main/java/io/druid/metadata/DefaultPasswordProvider.java]: We haven't done it for all files yet. Your suggestion to use the ASF style came after we had already updated the files. I figured we would update them going forward as we get more contributions, rather than update everything again. If people feel strongly about updating everything we can do that too.

File: common/src/main/java/io/druid/metadata/DefaultPasswordProvider.java (hunk 1 above)
Comment (xvrl) [inline common/src/main/java/io/druid/metadata/DefaultPasswordProvider.java]: @cheddar We shouldn't have to do the Map<> parsing if we let jackson handle the serde directly. All we need to do is change the defaultImpl to a separate class that has a JsonCreator method with a single string parameter (to support simple string), and leave everything else to use property annotations in the constructor (to handle the object style parsing)

**Comment by fjy** [comment]
Can we have some documentation about how the community can write their own extension or if they don't want to store their password in plain text in the config file?

**Comment by himanshug** [comment]
@fjy I wrote a quick reference implementation of a druid extension that shows how to implement your own password provider. we can add it in the documentation somewhere.
https://github.com/himanshug/druid-pwd-provider-extn-sample

**Comment by himanshug** [comment]
@xvrl @cheddar removed explicit map in DefaultPasswordProvider , it is still working fine.

**Comment by himanshug** [comment]
can someone pls restart the pull request build? it seems to have failed for some transient error? thanks.

**Comment by drcrallen** [comment]
@himanshug : close then open the PR

**Comment by himanshug** [comment]
need to close and reopen pull request so that build reruns.

File: common/src/test/java/io/druid/metadata/DefaultPasswordProviderTest.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  private static final String pwd = "nothing";
+
+  @Test
+  public void testFromStringPassword() {
+    DefaultPasswordProvider pp = new DefaultPasswordProvider(pwd);
+    Assert.assertEquals(pwd, pp.getPassword());
+  }
+  
+  @Test
+  public void testFromMapPassword() {
```
Comment (nishantmonu51) [inline common/src/test/java/io/druid/metadata/DefaultPasswordProviderTest.java]: rename this test method ?

File: common/src/test/java/io/druid/metadata/DefaultPasswordProviderTest.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    DefaultPasswordProvider pp = DefaultPasswordProvider.fromString(pwd);
+    Assert.assertEquals(pwd, pp.getPassword());
+  }
+}
```
Comment (nishantmonu51) [inline common/src/test/java/io/druid/metadata/DefaultPasswordProviderTest.java]: add unit test for json serialization/deserialization

Comment (himanshug) [inline common/src/test/java/io/druid/metadata/DefaultPasswordProviderTest.java]: json ser/de is being tested in 
MetadataStorageConnectorConfigTest.java and
MetadataStorageUpdaterJobSpecTest.java

but i'll add another one here as well. thanks.

**Comment by xvrl** [comment]
looks good to me, just squash your commits and update the license header so I can merge.

File: common/src/main/java/io/druid/metadata/MetadataStorageConnectorConfig.java [hunk 5]
```diff
@@ -38,8 +38,8 @@
   @JsonProperty
   private String user = null;
 
-  @JsonProperty
-  private String password = null;
+  @JsonProperty("password")
```
Comment (drcrallen) [inline common/src/main/java/io/druid/metadata/MetadataStorageConnectorConfig.java]: Do the docs need updated at all?

File: common/src/test/java/io/druid/metadata/DefaultPasswordProviderTest.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  @Test
+  public void testFromStringConstruction() {
+    DefaultPasswordProvider pp = DefaultPasswordProvider.fromString(pwd);
+    Assert.assertEquals(pwd, pp.getPassword());
+  }
+
+  @Test
+  public void testDeserializationFromJson() throws Exception {
+    PasswordProvider pp = jsonMapper.readValue(
+        "{\"type\": \"default\", \"password\": \"" + pwd + "\"}",
```
Comment (drcrallen) [inline common/src/test/java/io/druid/metadata/DefaultPasswordProviderTest.java]: Is the old behavior still supported? where simply passing in a string is taken as a string password?

File: common/src/main/java/io/druid/metadata/DefaultPasswordProvider.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+  @Override
+  @JsonProperty
+  public String getPassword()
+  {
+    return password;
+  }
+
+  @Override
+  public String toString() {
```
Comment (drcrallen) [inline common/src/main/java/io/druid/metadata/DefaultPasswordProvider.java]: General inquiry: Is there a way we can do a salt/hash or something so that the password in the log can be compared to the one used here?

**Comment by drcrallen** [comment]
This is a very good step forward towards better data security.

File: common/src/test/java/io/druid/metadata/DefaultPasswordProviderTest.java (hunk 6 above)
Comment (drcrallen) [inline common/src/test/java/io/druid/metadata/DefaultPasswordProviderTest.java]: I looked through the other tests and I'm still not sure if it is supposed to be supported. Would you mind putting in a test for the "legacy" case of cleartext passwords?

File: common/src/main/java/io/druid/metadata/DefaultPasswordProvider.java (hunk 7 above)
Comment (xvrl) [inline common/src/main/java/io/druid/metadata/DefaultPasswordProvider.java]: we currently just print out '*****' regardless of the size of the password, I don't think there is a need to print out anything special. Even printing out a salt/hash could be a security problem.

File: common/src/main/java/io/druid/metadata/MetadataStorageConnectorConfig.java (hunk 5 above)
Comment (xvrl) [inline common/src/main/java/io/druid/metadata/MetadataStorageConnectorConfig.java]: This is completely transparent to the user. We could add docs about the fact that there are password providers, but that might be more confusing than anything, given that we don't ship any other ones, I don't think this is necessary until we offer other password providers.

File: common/src/test/java/io/druid/metadata/DefaultPasswordProviderTest.java (hunk 6 above)
Comment (himanshug) [inline common/src/test/java/io/druid/metadata/DefaultPasswordProviderTest.java]: @drcallen yes, old behavior is fully supported and that is covered in the added unit tests for

common/src/test/java/io/druid/metadata/MetadataStorageConnectorConfigTest.java and
indexing-hadoop/src/test/java/io/druid/indexer/updater/MetadataStorageUpdaterJobSpecTest.java

however, I'll add additional test in DefaultPasswordProviderTest tool

**Comment by himanshug** [comment]
@xvrl license headers have been updated.
btw, squashing will change commit history.. does that mean I should close this and open another pull request with squashed commits?

File: common/src/main/java/io/druid/metadata/MetadataStorageConnectorConfig.java (hunk 5 above)
Comment (himanshug) [inline common/src/main/java/io/druid/metadata/MetadataStorageConnectorConfig.java]: may be we can add it to the list of [extendable] things at http://druid.io/docs/latest/Modules.html

**Comment by xvrl** [comment]
You can squash in your current branch, PR commit history should not matter to anyone until it has been merged into our working branches. We rewrite PR commit history all the time before merging. I wrote up instructions on squashing here if that's helpful: https://github.com/druid-io/druid/blob/master/CONTRIBUTING.md

File: common/src/main/java/io/druid/metadata/MetadataStorageConnectorConfig.java (hunk 5 above)
Comment (xvrl) [inline common/src/main/java/io/druid/metadata/MetadataStorageConnectorConfig.java]: Yes, there are many other things we need to add to that list, we can do that in a separate PR.

**Comment by himanshug** [comment]
@xvrl done

File: NOTICE [hunk 8]
```diff
@@ -1,5 +1,6 @@
 Druid - a distributed column store.
 Copyright 2012-2015 Metamarkets Group Inc.
+Copyright 2015 Yahoo! Inc.
```
Comment (xvrl) [inline NOTICE]: I am not a lawyer, but it might be confusing to include several copyright notices without context. I had a look at how it is done by a couple ASF projects, and it might be clearer to include the following snippet (based on the [Apache Derby notice](https://github.com/apache/derby/blob/trunk/NOTICE#L19)).

```
Portions of this software were originally developed by
Yahoo! Inc and are licensed to Metamarkets Group Inc. under the
"Metamarkets Corporate Contributor License Agreement"

The following copyright notice(s) were affixed to portions of the code
with which this file is now or was at one time distributed
and are placed here unaltered.

Copyright 2015 Yahoo! Inc.
```

Comment (himanshug) [inline NOTICE]: @xvrl thanks for the suggestion, but, wrt to licensing I'm just copy pasting stuff from internal email threads and update to NOTICE is found to be one acceptable way of doing it. but, anyway, let me ask @cheddar if he can get it resolved for us.

@cheddar can u pls take a look into this?

Comment (xvrl) [inline NOTICE]: I'll check with our counsel and we can update later, I'm fine leaving as is for now.

**Comment by xvrl** [comment]
TravisCI is undergoing maintenance. Tests pass fine locally for https://github.com/himanshug/druid/commit/126262edce6a3a4c7ca94fcf02f2ea229462b74c so I'm going to merge
